New Example (Swap to ZETA) and update to toolkit v8

fadeev opened the pull request:
* New example: Swap to ZETA
* Updated to toolkit v8-rc3, depends on https://github.com/zeta-chain/toolkit/pull/133
* Refactored Multioutput to avoid `CompilerError: Stack too deep, try removing local variables` error
* Replaced `require` with `revert` in the NFT example.

## Swap to ZETA

If the destination token address is WZETA, swap it to WZETA and don't withdraw.

Keeping both `Swap` and `SwapExtended` in the contracts directory. The thinking is, we keep the swap tutorial as is and introduce part 2 where we extend the swap to add "swap to ZETA" capability. Advantages: we still have a simple tutorial, and we offer more content for users who want to go further. Depends on https://github.com/zeta-chain/toolkit/pull/130

Can be simplified further: https://github.com/zeta-chain/toolkit/pull/129

Depends on https://github.com/zeta-chain/template/pull/55

fadeev commented:
Tested SwapExtended:

```
npx hardhat deploy --network zeta_testnet                    
🔑 Using account: 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32

🚀 Successfully deployed contract on zeta_testnet.
📜 Contract address: 0x68DFB7aF2037f5a3B05bcc490FCf8df10b4c6175
🌍 ZetaScan: https://athens.explorer.zetachain.com/address/0x68DFB7aF2037f5a3B05bcc490FCf8df10b4c6175
🌍 Blockcsout: https://zetachain-athens-3.blockscout.com/address/0x68DFB7aF2037f5a3B05bcc490FCf8df10b4c6175
```

```
npx hardhat interact --contract 0x68DFB7aF2037f5a3B05bcc490FCf8df10b4c6175 --network bsc_testnet --amount 0.01 --target-token 0x05BA149A7bd6dC1F937fA9046A9e05C05f3b18b0 --recipient 0x4955a3F38ff86ae92A914445099caa8eA2B9bA32
```

```
npx hardhat cctx 0xf10a3df6611f7b4d114f282491d7c1588ddb3d308ed3bea0fad4ac8a78a65ed6
^C✓ CCTXs on ZetaChain foun

✓ 0x39c6bf8727735e4a467dae5e98941539b549e411bb0b665a88a78dc8553a34bf: 97 → 7001: OutboundMined (Remote omni
chain contract call completed)
⠸ 0x1cf3d161e96f4e90cfacb7426fa6c7b31ea2fe70494adbd6747377b0f572d81a: 7001 → 11155111 (1 in queue): Pending
Outbound
```

fadeev commented:
Updated to Toolkit v8-rc1, which includes swap helper lib changes (accept system contract as arg) and changes to the deploy script, which allows choosing which contract to deploy.

fadeev commented:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.7;

import "@zetachain/protocol-contracts/contracts/zevm/SystemContract.sol";
import "@zetachain/protocol-contracts/contracts/zevm/interfaces/zContract.sol";
import "@zetachain/toolkit/contracts/SwapHelperLib.sol";
import "@zetachain/toolkit/contracts/BytesHelperLib.sol";
import "@zetachain/protocol-contracts/contracts/zevm/interfaces/IWZETA.sol";

contract SwapToZeta is zContract {
    SystemContract public systemContract;

    uint256 constant BITCOIN = 18332;

    constructor(address systemContractAddress) {
        systemContract = SystemContract(systemContractAddress);
    }

    modifier onlySystem() {
        require(
            msg.sender == address(systemContract),
            "Only system contract can call this function"
        );
        _;
    }

    function onCrossChainCall(
        zContext calldata context,
        address zrc20,
        uint256 amount,
        bytes calldata message
    ) external virtual override onlySystem {
        address target;
        bytes memory to;

        if (context.chainID == BITCOIN) {
            target = BytesHelperLib.bytesToAddress(message, 0);
            to = abi.encodePacked(BytesHelperLib.bytesToAddress(message, 20));
        } else {
            (address targetToken, bytes memory recipient) = abi.decode(
                message,
                (address, bytes)
            );
            target = targetToken;
            to = recipient;
        }

        address wzeta = systemContract.wZetaContractAddress();
        bool isTargetZeta = target == wzeta;
        uint256 inputForGas;
        address gasZRC20;
        uint256 gasFee;

        if (!isTargetZeta) {
            (gasZRC20, gasFee) = IZRC20(target).withdrawGasFee();

            inputForGas = SwapHelperLib.swapTokensForExactTokens(
                systemContract,
                zrc20,
                gasFee,
                gasZRC20,
                amount
            );
        }

        uint256 outputAmount = SwapHelperLib.swapExactTokensForTokens(
            systemContract,
            zrc20,
            isTargetZeta ? amount : amount - inputForGas,
            target,
            0
        );

        if (isTargetZeta) {
            IWETH9(wzeta).approve(address(this), outputAmount);
            IWETH9(wzeta).withdraw(outputAmount);
            payable(address(uint160(bytes20(to)))).transfer(outputAmount);
        } else {
            IZRC20(gasZRC20).approve(target, gasFee);
            IZRC20(target).withdraw(to, outputAmount);
        }
    }
}
```

andresaiello commented on omnichain/swap/contracts/SwapToZeta.sol:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    modifier onlySystem() {
+        require(
```
in this case I understand it's consistent to use require, but in general we try to use custom errors. Can us use this pr to change both? this new one and the same modifier in the other file?

fadeev replied:
Something like this?

```solidity
    error OnlySystemContract(string);

    modifier onlySystem() {
        if (msg.sender != address(systemContract)) {
            revert OnlySystemContract(
                "Only system contract can call this function"
            );
        }
        _;
    }
```

Instead of having this in the template, how about we make the modifier a part of another contract and import it?

fadeev replied:
If we don't want to include it in the `zContract`, maybe like this?

```solidity
contract Swap is zContract, OnlySystem {
  ...
}
```

@andresaiello

andresaiello replied:
even better

fadeev replied:
@andresaiello do we want to keep the `OnlySystem` modified in protocol contracts or toolkit? Probably toolkit, right?

fadeev replied:
https://github.com/zeta-chain/toolkit/pull/131

fadeev replied:
Let's merge and ship this in v8 and then I'll remove the modifier from example contracts.

fadeev commented:
* New example: Swap to ZETA
* Updated to toolkit v8-rc3, depends on https://github.com/zeta-chain/toolkit/pull/133
* Refactored Multioutput to avoid `CompilerError: Stack too deep, try removing local variables` error
* Replaced `require` with `revert` in the NFT example.
